uplc-syntax.md

```diff
@@ -34,12 +34,30 @@ module UPLC-CONCRETE-SYNTAX
                         | "bytestring"
                         | "unit"
                         | "bool"
+                        | "list" "(" TypeConstant ")"
+                        | "pair" "(" TypeConstant ")" "(" TypeConstant ")"
 
```

Comment:
You could also have `"pair" "(" TypeConstant "," TypeConstant ")"` here, and that would be more consistent with the notation we've used for types of built-in functions in the specification (although that's not syntax, but purely notation that we need to talk about signatures).

I think that "officially" we'd represent the type of (integer, string) pairs as
```
   [[(con pair) (con integer)] (con string)]
```
using application at the type level to say what the component types are. This is consistent with how complex types are represented internally and the prettyprinter may print types like this.  However, we haven't actually specified this in the current version of the specification yet (there's no way to write type applications in the grammar for untyped Plutus Core), so I think the suggestion here is perfectly sensible, and a lot more readable than the application syntax.  If the difference turns out to be a problem later on then I think we could allow the syntax here as sugar for the application syntax (which we might want to add later for compatibility with the prettyprinter), so I don't think we need to change it at the moment.

Reply:
Oh wait.  Instead of `(con list (integer)) you should maybe use `(con list (con integer))`. That's a bit annoying, but everywhere else type names appear inside `(con ...)`.  This is because in typed Plutus Core there are various non-built-in types and type variables,  and we have to distinguish (for example) the built-in type `(con integer)` from the type variable `integer` which is just as valid as `t` or `x`; it would obviously be silly to call a type variable `integer`, but writing code specifically to exclude built-in types in certain places would complicate matters a lot.